client/src/scenes/GameScene.ts

```diff
@@ -0,0 +1,211 @@
+        direction = Direction.Up;
+      } else if (keys.S.isDown) {
+        direction = Direction.Down;
+      } else if (keys.D.isDown) {
+        direction = Direction.Right;
+      } else if (keys.A.isDown) {
+        direction = Direction.Left;
+      } else {
+        direction = Direction.None;
+      }
+
+      if (prevDirection !== direction) {
+        // If connection is open and direction has changed, send updated direction
+        prevDirection = direction;
+        this.connection.sendMessage({ type: ClientMessageType.SetDirection, direction });
+      }
+    };
+
+    this.input.keyboard.on("keydown", handleKeyEvt);
+    this.input.keyboard.on("keyup", handleKeyEvt);
+
+    // Handle mouse-click input
+    this.input.on(Phaser.Input.Events.POINTER_DOWN, () => {
+      // If the connection is open, send through click events
+      this.connection.sendMessage({ type: ClientMessageType.Shoot });
+    });
+
+    // Render map objects
+    const mapGfx = this.add.graphics();
+
+    // Loop through each object in the map array...
+    MAP.forEach(({ x, y, width, height, color }) => {
+      // And draw the box according to it's properties
+      mapGfx.fillStyle(color, 1);
+      mapGfx.fillRect(x, y, width, height);
+    });
+
+    // Set the main camera's background colour
+    this.cameras.main.setBackgroundColor(BG_COLOR);
+  }
+
+  update() {
+    if (this.stateBuffer === undefined) {
+      return;
+    }
+
+    const { mousePointer } = this.input;
+    const { state } = this.stateBuffer.getInterpolatedState(Date.now());
+
+    this.syncSprites(
+      this.players,
+      new Map(
+        state.players.map((player) => [
+          player.id,
+          new Phaser.GameObjects.Sprite(this, player.position.x, player.position.y, "player").setRotation(
+            player.aimAngle
+          ),
+        ])
+      )
+    );
+
+    this.syncSprites(
+      this.bullets,
+      new Map(
+        state.bullets.map((bullet) => [
+          bullet.id,
+          new Phaser.GameObjects.Sprite(this, bullet.position.x, bullet.position.y, "bullet"),
+        ])
+      )
+    );
+
+    if (this.playerSprite) {
+      this.sendMousePosition(mousePointer, this.playerSprite);
+    }
+  }
+
+  private sendMousePosition(mousePointer: Phaser.Input.Pointer, playerSprite: Phaser.GameObjects.Sprite) {
+    const { x: mouseX, y: mouseY } = mousePointer;
+    const { x: playerX, y: playerY } = playerSprite;
+    const { zoom, worldView } = this.cameras.main;
+
+    // Establish the angle between the player's camera-relative position and the mouse
+    const relX = (playerX - worldView.x) * zoom;
+    const relY = (playerY - worldView.y) * zoom;
+    const aimRad = pMath.Angle.Between(relX + zoom, relY + zoom, mouseX, mouseY);
+    const aimMoved = this.prevAimRad !== aimRad;
+
+    // Only if the aim has updated, send the update
+    if (aimMoved) {
+      this.connection.sendMessage({ type: ClientMessageType.SetAngle, angle: aimRad });
```

Comment:
if you open up the network tab and look at the websocket messages, you'll see that we're sending a ton of packets every second, even if the character or mouse aren't moving. can we double check this logic to make sure we're not sending duplicate updates?

Reply:
found the bug, you can fix it like this:
```ts
    const aimRad = pMath.Angle.Between(relX + zoom, relY + zoom, mouseX, mouseY);

    // Only if the aim has updated, send the update
    if (this.prevAimRad !== aimRad) {
      this.connection.sendMessage({ type: ClientMessageType.SetAngle, angle: aimRad });
      this.prevAimRad = aimRad;
    }
```

Reply:
Awesome, thanks! 🙌